File: src/main/kotlin/inflearn/java/ch01_string/_1_10_가장_짧은_문자거리.java
package inflearn.java.ch01_string;

import java.util.Arrays;
import java.util.Scanner;
import java.util.StringTokenizer;

public class _1_10_가장_짧은_문자거리 {
    public static int[] solution(String s, char t) {
        int[] answer = new int[s.length()];
        int p = 1000;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == t) {
                p = 0;
            } else {
                p += 1;
            }
            answer[i] = p;
        }
        p = 1000;

        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) == t) {
                p = 0;
            } else {
                p += 1;
                answer[i] = Math.min(answer[i], p);
            }
        }

        return answer;
    }

    public static String solutionTRESH(String s, char t) {
        int[] memo = new int[s.length()];
        Arrays.fill(memo, Integer.MAX_VALUE);

        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                if (s.charAt(j) == t) {
                    memo[i] = j - i;
                    break;
                }
            }
            for (int j = i; j >= 0; j--) {
                if (s.charAt(j) == t) {
                    if (memo[i] < i - j)
                        break;
                    memo[i] = i - j;
                    break;
                }
            }
        }

        StringBuilder answer = new StringBuilder();

        for (int dis : memo) {
            answer.append(dis).append(" ");
        }

        return answer.toString();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        StringTokenizer st = new StringTokenizer(sc.nextLine());
        for (int anInt : solution(st.nextToken(), st.nextToken().charAt(0))) {
            System.out.print(anInt);
            System.out.print(" ");
        }
    }
}

File: src/main/kotlin/inflearn/1_10 가장 짧은 문자거리.kt
package inflearn

import java.util.*

object `1_10 가장 짧은 문자거리` {
    fun solution(s: String, t: Char): IntArray {
        val answer = IntArray(s.length)
        var p = 1000
        for (i in s.indices) {
            if (s[i] == t) {
                p = 0
            } else {
                p += 1
            }
            answer[i] = p
        }
        p = 1000
        for (i in s.length - 1 downTo 0) {
            if (s[i] == t) {
                p = 0
            } else {
                p += 1
                answer[i] = Math.min(answer[i], p)
            }
        }
        return answer
    }

    fun solutionTRESH(s: String, t: Char): String? {
        val memo = IntArray(s.length)
        Arrays.fill(memo, Int.MAX_VALUE)
        for (i in s.indices) {
            for (j in i until s.length) {
                if (s[j] == t) {
                    memo[i] = j - i
                    break
                }
            }
            for (j in i downTo 0) {
                if (s[j] == t) {
                    if (memo[i] < i - j) break
                    memo[i] = i - j
                    break
                }
            }
        }
        val answer = StringBuilder()
        for (dis in memo) {
            answer.append(dis).append(" ")
        }
        return answer.toString()
    }

    @JvmStatic
    fun main(args: Array<String>) {
        val sc = Scanner(System.`in`)
        val st = StringTokenizer(sc.nextLine())
        for (anInt in solution(st.nextToken(), st.nextToken()[0])) {
            print(anInt)
            print(" ")
        }
    }
}